nitro-protocol/contracts/ForceMove.sol

```diff
@@ -451,132 +467,85 @@ contract ForceMove is IForceMove, StatusManager {
      * @return The hash of the latest state in the proof, if supported, else reverts.
      */
     function _requireStateSupportedBy(
-        IForceMoveApp.VariablePart[] memory variableParts,
-        bytes32 channelId,
         FixedPart memory fixedPart,
+        VariablePart[] memory variableParts,
+        bytes32 channelId,
         Signature[] memory sigs,
         uint8[] memory whoSignedWhat
     ) internal pure returns (bytes32) {
-        bytes32[] memory stateHashes = _requireValidTransitionChain(
-            variableParts,
-            channelId,
-            fixedPart
-        );
+        VariablePart memory latestVariablePart = IForceMoveApp(fixedPart.appDefinition)
+            .latestSupportedState(fixedPart, _bindSigsToVariableParts(variableParts, sigs, whoSignedWhat));
 
-        require(
-            _validSignatures(
-                _lastVariablePart(variableParts).turnNum,
-                fixedPart.participants,
-                stateHashes,
-                sigs,
-                whoSignedWhat
-            ),
-            'Invalid signatures'
-        );
+        // enforcing the latest supported state being in the last slot of the array
+        _requireVariablePartIsLast(latestVariablePart, variableParts);
```

Comment:
This needs to be enforced explicitly, otherwise undesired behavior could appear.

Let's take [`challenge` method](https://github.com/statechannels/go-nitro/blob/integrate-latest-supported-state/nitro-protocol/contracts/ForceMove.sol#L45) for example:

Firstly we look up `largestTurnNum` **of the last supplied variable part**
``` solidity
uint48 largestTurnNum = _lastVariablePart(signedVariableParts).turnNum;
```
for future checks
``` solidity
if (_mode(channelId) == ChannelMode.Open) {
    // using largestTurnNum of the last VariablePart
    _requireNonDecreasedTurnNumber(channelId, largestTurnNum);
} else if (_mode(channelId) == ChannelMode.Challenge) {
    // using largestTurnNum of the last VariablePart
    _requireIncreasedTurnNumber(channelId, largestTurnNum);
} else {
    _requireChannelNotFinalized(channelId);
}
```
Thus, we _assume_ that `_requireStateSupportedBy` (and therefore `latestSupportedState`) perceive last variable part as latest supported one.

Suppose developers described such logic, that _`requireStateSupportedBy` returned the hash **of not the last state**. Then, by the means of dev's DApp, not the last state should be marked as challenging, but that is not what happens:
``` solidity
statusOf[channelId] = _generateStatus(
        //  \/ data from the last VariablePart \/
        largestTurnNum,
        uint48(block.timestamp) + fixedPart.challengeDuration,
        //  \/ data from NOT the last VariablePart  \/
        supportedStateHash,
        //  \/ data from the last VariablePart \/
        _hashOutcome(_lastVariablePart(signedVariableParts).outcome)
    )
);
```
Parts of the last VariablePart alongside with hash of NOT the last state are saved, which will lead to behavior errors.

Described undesired logic can be easily fixed by only using the data gained from `_requireStateSupportedBy`, however that means we are allowing dev's DApp to return not only the last state, which is the opposite of what we are trying to enforce.

Having said that, the only way I see we enforce DApp returning only the last state is by explicitly requiring so.

Reply:
You're right about this. However, I am curious what you think about this approach #666. It may go against what I said before 🤦 but I think it might help reduce confusion around "last" (`arr.length-1`) and "latest" (greatest turn number).

Reply:
#666 suggests `latestSupportedState` can return `VariablePart` with not the greatest turn number amongst what was passed to this method. This can happen in a situation, where first N `VariableParts` constitutes a support proof, but the last M do not fall under the rules of DApp, thus are not in support proof. Let's call this situation "sub-support proof". So latest `VariablePart` in provided array (expected support proof) will not be equal to latest one in sub-support proof.

Therefore, confusions can emerge. Imagine participants want to conclude a channel and pass a support proof to `latestSupportedState`, expecting the channel to conclude with outcome specified in the latest `VariablePart`. However, by the logic specified in DApp, latest `VariablePart` appears not to be in support proof (either because participants made a mistake and DApp returned sub-support proof or DApp is malicious), consequently, not the latest `VariablePart` is returned and channel concludes with outdated outcome.

This means, that by not explicitly requiring `latestSupportedState` to return the latest `VariablePart`, we shift responsibility of ensuring channel will conclude / make a checkpoint / challenge with a desired (latest provided) `VariablePart` from `ForceMove` to the participants, who are prone to errors. And consequently, another possible _vector of attack / ability to make a mistake_ appears.

Or we can lay this responsibility on `ForceMove` by requiring `signedVariableParts` array to be increasingly sorted by turn number and `latestSupportedState` to return the last (thus, expected latest) `VariablePart`.

Reply:
Logical continuation of this thread is [#764](https://github.com/statechannels/go-nitro/discussions/764).